api/plugin.go

```diff
@@ -0,0 +1,355 @@
+//   200:
+//     description: Ok
+//     schema:
+//         type: object
+//         description: plugin configuration
+//   404:
+//     description: Not Found
+//     schema:
+//         $ref: "#/definitions/Error"
+//   500:
+//     description: Internal Server Error
+//     schema:
+//         $ref: "#/definitions/Error"
+//   501:
+//     description: Not Implemented
+//     schema:
+//         $ref: "#/definitions/Error"
+func (c *PluginAPI) GetConfig(ctx *gin.Context) {
+	withID(ctx, "id", func(id uint) {
+		plugin := c.DB.GetPluginConfByID(id)
+		if plugin == nil {
+			ctx.AbortWithError(404, errors.New("unknown plugin"))
+			return
+		}
+		instance, err := c.Manager.Instance(id)
+		if err != nil {
+			ctx.AbortWithError(500, errors.New("plugin instance not found"))
+			return
+		}
+
+		if aborted := supportOrAbort(ctx, instance, compat.Configurer); aborted {
+			return
+		}
+
+		ctx.Header("content-type", "application/x-yaml")
+		ctx.Writer.Write(plugin.Config)
+	})
+
+}
+
+// UpdateConfig updates Configurer plugin configuration in YAML format.
+// swagger:operation POST /plugin/:id/config plugin updatePluginConfig
+//
+// Update YAML configuration for Configurer plugin.
+//
+// ---
+// consumes: [application/x-yaml]
+// produces: [application/json]
+// security: [clientTokenHeader: [], clientTokenQuery: [], basicAuth: []]
+// responses:
+//   200:
+//     description: Ok
+//   400:
+//     description: Bad Request
+//     schema:
+//         $ref: "#/definitions/Error"
+//   404:
+//     description: Not Found
+//     schema:
+//         $ref: "#/definitions/Error"
+//   500:
+//     description: Internal Server Error
+//     schema:
+//         $ref: "#/definitions/Error"
+//   501:
+//     description: Not Implemented
+//     schema:
+//         $ref: "#/definitions/Error"
+func (c *PluginAPI) UpdateConfig(ctx *gin.Context) {
+	withID(ctx, "id", func(id uint) {
+		conf := c.DB.GetPluginConfByID(id)
+		if conf == nil {
+			ctx.AbortWithError(404, errors.New("unknown plugin"))
+			return
+		}
+		instance, err := c.Manager.Instance(id)
+		if err != nil {
+			ctx.AbortWithError(500, errors.New("plugin instance not found"))
+			return
+		}
+
+		if aborted := supportOrAbort(ctx, instance, compat.Configurer); aborted {
+			return
+		}
+
+		newConf := instance.EmptyConfig()
+		newconfBytes, err := ioutil.ReadAll(ctx.Request.Body)
+		if err != nil {
+			ctx.AbortWithError(500, err)
+			return
+		}
+		if err := yaml.Unmarshal(newconfBytes, newConf); err != nil {
+			ctx.AbortWithError(400, err)
+			return
+		}
+		if err := instance.ValidateAndSetConfig(newConf); err != nil {
+			ctx.AbortWithError(400, err)
+			return
+		}
+		conf.Config = newconfBytes
+		c.DB.UpdatePluginConf(conf)
+	})
+}
+
+func supportOrAbort(ctx *gin.Context, instance compat.PluginInstance, module compat.Module) (aborted bool) {
+	if compat.HasSupport(instance, module) {
+		return false
+	}
+	ctx.AbortWithError(400, fmt.Errorf("plugin does not support %s", module))
```

Comment:
Consider use a `501 Not Implemented` here

Reply:
We provide a capabilities api, in there it is specified which apis are supported for a given plugin. Therefore I chose 400 because the client made a mistake.

Also, the rfc states that 'not implemented' means that it is unsupported for any resource (in this case plugins) and this is not true:
> The 501 (Not Implemented) status code indicates that the server does
   not support the functionality required to fulfill the request.  This
   is the appropriate response when the server does not recognize the
   request method and is not capable of supporting it for any resource.

https://tools.ietf.org/html/rfc7231#section-6.6.2

Reply:
Okay, I will update the swagger docs for this